File: step_counter.js
#!/usr/bin/env node
// step_counter.js
const { program } = require('commander');
const fs = require('fs');
const chalk = require('chalk');

const DATA_FILE = 'steps.json';
const DEFAULT_GOAL = 10000;

class StepCounter {
    constructor() {
        this.data = this.load();
        this.goal = this.data.goal || DEFAULT_GOAL;
        this.history = this.data.history || [];
    }

    load() {
        try {
            if (fs.existsSync(DATA_FILE)) {
                return JSON.parse(fs.readFileSync(DATA_FILE, 'utf8'));
            }
        } catch (e) {}
        return { goal: DEFAULT_GOAL, history: [] };
    }

    save() {
        fs.writeFileSync(DATA_FILE, JSON.stringify({ goal: this.goal, history: this.history }, null, 2));
    }

    getToday() {
        const today = new Date().toISOString().split('T')[0];
        return this.history.find(e => e.date === today);
    }

    addSteps(steps) {
        const today = new Date().toISOString().split('T')[0];
        const entry = this.getToday();
        if (entry) {
            entry.steps += steps;
        } else {
            this.history.push({ date: today, steps });
        }
        this.save();
        const total = entry ? entry.steps : steps;
        console.log(chalk.green(`Добавлено ${steps} шагов. Всего сегодня: ${total}`));
    }

    show() {
        const today = new Date().toISOString().split('T')[0];
        const entry = this.getToday();
        const stepsToday = entry ? entry.steps : 0;
        const progress = Math.min(100, Math.floor((stepsToday / this.goal) * 100));
        const barLen = 20;
        const filled = Math.floor(barLen * progress / 100);
        const bar = '█'.repeat(filled) + '░'.repeat(barLen - filled);
        console.log(chalk.cyan(`📊 Сегодня: ${stepsToday} шагов`));
        console.log(chalk.yellow(`Цель: ${this.goal} шагов`));
        console.log(`Прогресс: ${chalk.green(bar)} ${progress}%`);
        console.log(chalk.magenta('\nИстория за 7 дней:'));
        const todayDt = new Date();
        for (let i = 6; i >= 0; i--) {
            const d = new Date(todayDt);
            d.setDate(d.getDate() - i);
            const dateStr = d.toISOString().split('T')[0];
            const found = this.history.find(e => e.date === dateStr);
            console.log(`  ${dateStr}: ${found ? found.steps : 0}`);
        }
    }

    setGoal(goal) {
        this.goal = goal;
        this.save();
        console.log(chalk.green(`Цель установлена: ${goal} шагов`));
    }

    history() {
        console.log(chalk.cyan('История (все записи):'));
        this.history.sort((a, b) => a.date.localeCompare(b.date));
        for (const e of this.history) {
            console.log(`  ${e.date}: ${e.steps}`);
        }
    }

    reset() {
        this.history = [];
        this.save();
        console.log(chalk.green('Все данные сброшены.'));
    }

    exportCsv(filename) {
        const header = 'date,steps\n';
        const rows = this.history.map(e => `${e.date},${e.steps}`).join('\n');
        fs.writeFileSync(filename, header + rows);
        console.log(chalk.green(`Экспортировано в ${filename} (CSV)`));
    }
}

program
    .option('--add <steps>', 'Добавить шаги за сегодня', parseInt)
    .option('--show', 'Показать сегодняшний прогресс и историю')
    .option('--goal <steps>', 'Установить дневную цель', parseInt)
    .option('--history', 'Показать всю историю')
    .option('--reset', 'Сбросить все данные')
    .option('--export-csv <file>', 'Экспорт истории в CSV')
    .parse(process.argv);

const opts = program.opts();
const counter = new StepCounter();

if (opts.add) counter.addSteps(opts.add);
else if (opts.show) counter.show();
else if (opts.goal) counter.setGoal(opts.goal);
else if (opts.history) counter.history();
else if (opts.reset) counter.reset();
else if (opts.exportCsv) counter.exportCsv(opts.exportCsv);
else program.help();

File: StepCounter.cs
// StepCounter.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace StepCounter
{
    class Program
    {
        static void Main(string[] args)
        {
            var opts = ParseArgs(args);
            var counter = new StepCounter();
            if (opts.Add.HasValue) counter.AddSteps(opts.Add.Value);
            else if (opts.Show) counter.Show();
            else if (opts.Goal.HasValue) counter.SetGoal(opts.Goal.Value);
            else if (opts.History) counter.History();
            else if (opts.Reset) counter.Reset();
            else if (opts.ExportCsv != null) counter.ExportCsv(opts.ExportCsv);
            else Console.WriteLine("Используйте --help для справки.");
        }

        static Options ParseArgs(string[] args)
        {
            var opts = new Options();
            for (int i = 0; i < args.Length; i++)
            {
                switch (args[i])
                {
                    case "--add": opts.Add = int.Parse(args[++i]); break;
                    case "--show": opts.Show = true; break;
                    case "--goal": opts.Goal = int.Parse(args[++i]); break;
                    case "--history": opts.History = true; break;
                    case "--reset": opts.Reset = true; break;
                    case "--export-csv": opts.ExportCsv = args[++i]; break;
                }
            }
            return opts;
        }

        class Options
        {
            public int? Add { get; set; }
            public bool Show { get; set; }
            public int? Goal { get; set; }
            public bool History { get; set; }
            public bool Reset { get; set; }
            public string ExportCsv { get; set; }
        }

        class HistoryEntry
        {
            public string Date { get; set; }
            public int Steps { get; set; }
        }

        class Data
        {
            public int Goal { get; set; } = 10000;
            public List<HistoryEntry> History { get; set; } = new List<HistoryEntry>();
        }

        class StepCounter
        {
            private const string DataFile = "steps.json";
            private Data data;

            public StepCounter()
            {
                Load();
            }

            private void Load()
            {
                try
                {
                    if (File.Exists(DataFile))
                    {
                        string json = File.ReadAllText(DataFile);
                        data = JsonSerializer.Deserialize<Data>(json) ?? new Data();
                    }
                    else
                        data = new Data();
                }
                catch
                {
                    data = new Data();
                }
                if (data.History == null) data.History = new List<HistoryEntry>();
            }

            private void Save()
            {
                string json = JsonSerializer.Serialize(data, new JsonSerializerOptions { WriteIndented = true });
                File.WriteAllText(DataFile, json);
            }

            private HistoryEntry GetToday()
            {
                string today = DateTime.UtcNow.ToString("yyyy-MM-dd");
                return data.History.FirstOrDefault(e => e.Date == today);
            }

            public void AddSteps(int steps)
            {
                var entry = GetToday();
                if (entry != null)
                    entry.Steps += steps;
                else
                    data.History.Add(new HistoryEntry { Date = DateTime.UtcNow.ToString("yyyy-MM-dd"), Steps = steps });
                Save();
                int total = entry != null ? entry.Steps : steps;
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"Добавлено {steps} шагов. Всего сегодня: {total}");
                Console.ResetColor();
            }

            public void Show()
            {
                var entry = GetToday();
                int stepsToday = entry != null ? entry.Steps : 0;
                int progress = Math.Min(100, stepsToday * 100 / data.Goal);
                int barLen = 20;
                int filled = progress * barLen / 100;
                string bar = new string('█', filled) + new string('░', barLen - filled);
                Console.ForegroundColor = ConsoleColor.Cyan;
                Console.WriteLine($"📊 Сегодня: {stepsToday} шагов");
                Console.ForegroundColor = ConsoleColor.Yellow;
                Console.WriteLine($"Цель: {data.Goal} шагов");
                Console.ResetColor();
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"Прогресс: {bar} {progress}%");
                Console.ResetColor();
                Console.ForegroundColor = ConsoleColor.Magenta;
                Console.WriteLine("\nИстория за 7 дней:");
                Console.ResetColor();
                for (int i = 6; i >= 0; i--)
                {
                    var d = DateTime.UtcNow.AddDays(-i);
                    string dateStr = d.ToString("yyyy-MM-dd");
                    var found = data.History.FirstOrDefault(e => e.Date == dateStr);
                    Console.WriteLine($"  {dateStr}: {(found != null ? found.Steps : 0)}");
                }
            }

            public void SetGoal(int goal)
            {
                data.Goal = goal;
                Save();
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"Цель установлена: {goal} шагов");
                Console.ResetColor();
            }

            public void History()
            {
                Console.ForegroundColor = ConsoleColor.Cyan;
                Console.WriteLine("История (все записи):");
                Console.ResetColor();
                foreach (var e in data.History.OrderBy(e => e.Date))
                    Console.WriteLine($"  {e.Date}: {e.Steps}");
            }

            public void Reset()
            {
                data.History.Clear();
                Save();
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine("Все данные сброшены.");
                Console.ResetColor();
            }

            public void ExportCsv(string filename)
            {
                using var sw = new StreamWriter(filename);
                sw.WriteLine("date,steps");
                foreach (var e in data.History)
                    sw.WriteLine($"{e.Date},{e.Steps}");
                Console.ForegroundColor = ConsoleColor.Green;
                Console.WriteLine($"Экспортировано в {filename} (CSV)");
                Console.ResetColor();
            }
        }
    }
}
